File: app/src/main/java/com/example/crypto/DeviceE2ee.kt
package com.example.crypto

import android.security.keystore.KeyGenParameterSpec
import android.security.keystore.KeyProperties
import android.util.Base64
import java.security.KeyFactory
import java.security.KeyPair
import java.security.KeyPairGenerator
import java.security.KeyStore
import java.security.MessageDigest
import java.security.spec.MGF1ParameterSpec
import java.security.spec.X509EncodedKeySpec
import javax.crypto.Cipher
import javax.crypto.KeyGenerator
import javax.crypto.SecretKey
import javax.crypto.spec.GCMParameterSpec
import javax.crypto.spec.OAEPParameterSpec
import javax.crypto.spec.PSource
import javax.crypto.spec.SecretKeySpec

data class RecipientDeviceKey(
    val deviceId: String,
    val publicKeyBase64: String
)

data class E2eeEnvelope(
    val ciphertextBase64: String,
    val nonceBase64: String,
    val wrappedKeys: Map<String, String>,
    val protocolVersion: String = DeviceE2ee.PROTOCOL_VERSION
)

data class EncryptedAttachment(
    val ciphertext: ByteArray,
    val keyBase64: String,
    val nonceBase64: String
)

/**
 * ChatNU's production E2EE envelope.
 *
 * Each message uses a fresh AES-256-GCM content key. That content key is wrapped independently
 * for every active recipient device with that device's RSA-3072 OAEP public key. Private identity
 * keys are generated inside Android Keystore and never uploaded.
 *
 * This is real end-to-end encryption, but it is deliberately not described as Signal/Double
 * Ratchet: version 1 does not provide Signal-style per-message forward secrecy or deniability.
 */
class DeviceE2ee {
    fun publicKeyBase64(account: String): String {
        val pair = getOrCreateIdentity(account)
        return encode(pair.public.encoded)
    }

    fun encryptMessage(
        plaintext: ByteArray,
        recipientKeys: List<RecipientDeviceKey>,
        aad: String
    ): E2eeEnvelope {
        require(recipientKeys.isNotEmpty()) { "No recipient device keys are available" }
        val contentKey = randomAesKey()
        val cipher = Cipher.getInstance("AES/GCM/NoPadding")
        cipher.init(Cipher.ENCRYPT_MODE, contentKey)
        cipher.updateAAD(aad.toByteArray(Charsets.UTF_8))
        val ciphertext = cipher.doFinal(plaintext)

        val wrapped = linkedMapOf<String, String>()
        recipientKeys.distinctBy { it.deviceId }.forEach { recipient ->
            val remotePublicKey = KeyFactory.getInstance("RSA").generatePublic(
                X509EncodedKeySpec(decode(recipient.publicKeyBase64))
            )
            val wrapper = rsaOaepCipher()
            wrapper.init(Cipher.ENCRYPT_MODE, remotePublicKey, OAEP_SHA256_MGF1_SHA1)
            wrapped[recipient.deviceId] = encode(wrapper.doFinal(contentKey.encoded))
        }

        return E2eeEnvelope(
            ciphertextBase64 = encode(ciphertext),
            nonceBase64 = encode(cipher.iv),
            wrappedKeys = wrapped
        )
    }

    fun decryptMessage(
        account: String,
        deviceId: String,
        ciphertextBase64: String,
        nonceBase64: String,
        metadata: Map<String, Any?>?,
        aad: String
    ): ByteArray {
        val wrappedMap = metadata?.get("wrappedKeys") as? Map<*, *>
            ?: error("Encrypted message is missing wrappedKeys")
        val wrappedKey = wrappedMap.entries
            .firstOrNull { it.key?.toString() == deviceId }
            ?.value
            ?.toString()
            ?: error("This message was not encrypted for the current device")

        val privateKey = getOrCreateIdentity(account).private
        val unwrapper = rsaOaepCipher()
        unwrapper.init(Cipher.DECRYPT_MODE, privateKey, OAEP_SHA256_MGF1_SHA1)
        val contentKey = SecretKeySpec(unwrapper.doFinal(decode(wrappedKey)), "AES")

        val cipher = Cipher.getInstance("AES/GCM/NoPadding")
        cipher.init(
            Cipher.DECRYPT_MODE,
            contentKey,
            GCMParameterSpec(128, decode(nonceBase64))
        )
        cipher.updateAAD(aad.toByteArray(Charsets.UTF_8))
        return cipher.doFinal(decode(ciphertextBase64))
    }

    fun encryptAttachment(plaintext: ByteArray): EncryptedAttachment {
        val key = randomAesKey()
        val cipher = Cipher.getInstance("AES/GCM/NoPadding")
        cipher.init(Cipher.ENCRYPT_MODE, key)
        return EncryptedAttachment(
            ciphertext = cipher.doFinal(plaintext),
            keyBase64 = encode(key.encoded),
            nonceBase64 = encode(cipher.iv)
        )
    }

    fun decryptAttachment(ciphertext: ByteArray, keyBase64: String, nonceBase64: String): ByteArray {
        val cipher = Cipher.getInstance("AES/GCM/NoPadding")
        cipher.init(
            Cipher.DECRYPT_MODE,
            SecretKeySpec(decode(keyBase64), "AES"),
            GCMParameterSpec(128, decode(nonceBase64))
        )
        return cipher.doFinal(ciphertext)
    }

    private fun getOrCreateIdentity(account: String): KeyPair {
        val alias = identityAlias(account)
        val keyStore = KeyStore.getInstance("AndroidKeyStore").apply { load(null) }
        val existingPrivate = keyStore.getKey(alias, null)
        val existingPublic = keyStore.getCertificate(alias)?.publicKey
        if (existingPrivate != null && existingPublic != null) {
            return KeyPair(existingPublic, existingPrivate as java.security.PrivateKey)
        }

        val generator = KeyPairGenerator.getInstance(KeyProperties.KEY_ALGORITHM_RSA, "AndroidKeyStore")
        generator.initialize(
            KeyGenParameterSpec.Builder(
                alias,
                KeyProperties.PURPOSE_ENCRYPT or KeyProperties.PURPOSE_DECRYPT
            )
                .setKeySize(3072)
                .setDigests(KeyProperties.DIGEST_SHA256, KeyProperties.DIGEST_SHA1)
                .setEncryptionPaddings(KeyProperties.ENCRYPTION_PADDING_RSA_OAEP)
                .build()
        )
        return generator.generateKeyPair()
    }

    private fun identityAlias(account: String): String {
        val digest = MessageDigest.getInstance("SHA-256")
            .digest(account.trim().lowercase().toByteArray(Charsets.UTF_8))
            .joinToString("") { "%02x".format(it) }
            .take(24)
        return "chatnu_e2ee_identity_$digest"
    }

    private fun randomAesKey(): SecretKey = KeyGenerator.getInstance("AES").run {
        init(256)
        generateKey()
    }

    private fun rsaOaepCipher(): Cipher = Cipher.getInstance("RSA/ECB/OAEPWithSHA-256AndMGF1Padding")

    private fun encode(bytes: ByteArray): String = Base64.encodeToString(bytes, Base64.NO_WRAP)
    private fun decode(value: String): ByteArray = Base64.decode(value, Base64.NO_WRAP)

    companion object {
        const val PROTOCOL_VERSION = "chatnu-e2ee-rsa3072-aes256gcm-v1"

        private val OAEP_SHA256_MGF1_SHA1 = OAEPParameterSpec(
            "SHA-256",
            "MGF1",
            MGF1ParameterSpec.SHA1,
            PSource.PSpecified.DEFAULT
        )
    }
}

File: flutter/lib/core/crypto/device_e2ee.dart
import 'dart:convert';
import 'dart:typed_data';

import 'package:chatnu/core/crypto/portable_identity_store.dart';
import 'package:chatnu/core/crypto/rsa_der_codec.dart';
import 'package:chatnu/core/platform/chatnu_native_bridge.dart';
import 'package:pointycastle/export.dart';

class RecipientDeviceKey {
  const RecipientDeviceKey({
    required this.deviceId,
    required this.publicKeyBase64,
  });

  final String deviceId;
  final String publicKeyBase64;
}

class E2eeEnvelope {
  const E2eeEnvelope({
    required this.ciphertextBase64,
    required this.nonceBase64,
    required this.wrappedKeys,
    this.protocolVersion = DeviceE2ee.protocolVersion,
  });

  final String ciphertextBase64;
  final String nonceBase64;
  final Map<String, String> wrappedKeys;
  final String protocolVersion;
}

class EncryptedAttachment {
  const EncryptedAttachment({
    required this.ciphertext,
    required this.keyBase64,
    required this.nonceBase64,
  });

  final Uint8List ciphertext;
  final String keyBase64;
  final String nonceBase64;
}

class DeviceE2ee {
  DeviceE2ee({
    required PortableIdentityStore portableIdentities,
    required ChatNuNativeBridge nativeBridge,
  }) : _portableIdentities = portableIdentities,
       _nativeBridge = nativeBridge;

  static const protocolVersion = 'chatnu-e2ee-rsa3072-aes256gcm-v1';

  final PortableIdentityStore _portableIdentities;
  final ChatNuNativeBridge _nativeBridge;

  Future<String> publicKeyBase64(String account) async {
    final native = await _nativeBridge.publicIdentityKey(account);
    if (native != null && native.isNotEmpty) return native;
    final identity = await _portableIdentities.getOrCreate(account);
    return base64.encode(
      RsaDerCodec.encodeSubjectPublicKeyInfo(identity.publicKey),
    );
  }

  Future<E2eeEnvelope> encryptMessage({
    required Uint8List plaintext,
    required List<RecipientDeviceKey> recipientKeys,
    required String aad,
  }) async {
    if (recipientKeys.isEmpty) {
      throw StateError('No recipient device keys are available.');
    }
    final contentKey = PortableIdentityStore.secureRandom().nextBytes(32);
    final nonce = PortableIdentityStore.secureRandom().nextBytes(12);
    final ciphertext = _aesGcm(
      encrypt: true,
      data: plaintext,
      key: contentKey,
      nonce: nonce,
      aad: Uint8List.fromList(utf8.encode(aad)),
    );

    final wrappedKeys = <String, String>{};
    final seen = <String>{};
    for (final recipient in recipientKeys) {
      if (!seen.add(recipient.deviceId)) continue;
      final publicKey = RsaDerCodec.decodeSubjectPublicKeyInfo(
        base64.decode(recipient.publicKeyBase64),
      );
      wrappedKeys[recipient.deviceId] = base64.encode(
        _wrapContentKey(publicKey, contentKey),
      );
    }
    return E2eeEnvelope(
      ciphertextBase64: base64.encode(ciphertext),
      nonceBase64: base64.encode(nonce),
      wrappedKeys: wrappedKeys,
    );
  }

  Future<Uint8List> decryptMessage({
    required String account,
    required String deviceId,
    required String ciphertextBase64,
    required String nonceBase64,
    required Map<String, dynamic>? metadata,
    required String aad,
  }) async {
    final rawWrapped = metadata?['wrappedKeys'];
    if (rawWrapped is! Map) {
      throw StateError('Encrypted message is missing wrappedKeys.');
    }
    String? wrappedKey;
    for (final entry in rawWrapped.entries) {
      if (entry.key.toString() == deviceId) {
        wrappedKey = entry.value?.toString();
        break;
      }
    }
    if (wrappedKey == null || wrappedKey.isEmpty) {
      throw StateError(
        'This message was not encrypted for the current device.',
      );
    }

    Uint8List contentKey;
    final native = await _nativeBridge.unwrapContentKey(
      account: account,
      wrappedKeyBase64: wrappedKey,
    );
    if (native != null) {
      contentKey = Uint8List.fromList(native);
    } else {
      final identity = await _portableIdentities.getOrCreate(account);
      contentKey = _unwrapContentKey(
        identity.privateKey,
        Uint8List.fromList(base64.decode(wrappedKey)),
      );
    }

    return _aesGcm(
      encrypt: false,
      data: Uint8List.fromList(base64.decode(ciphertextBase64)),
      key: contentKey,
      nonce: Uint8List.fromList(base64.decode(nonceBase64)),
      aad: Uint8List.fromList(utf8.encode(aad)),
    );
  }

  EncryptedAttachment encryptAttachment(Uint8List plaintext) {
    final key = PortableIdentityStore.secureRandom().nextBytes(32);
    final nonce = PortableIdentityStore.secureRandom().nextBytes(12);
    return EncryptedAttachment(
      ciphertext: _aesGcm(
        encrypt: true,
        data: plaintext,
        key: key,
        nonce: nonce,
        aad: Uint8List(0),
      ),
      keyBase64: base64.encode(key),
      nonceBase64: base64.encode(nonce),
    );
  }

  Uint8List decryptAttachment({
    required Uint8List ciphertext,
    required String keyBase64,
    required String nonceBase64,
  }) {
    return _aesGcm(
      encrypt: false,
      data: ciphertext,
      key: Uint8List.fromList(base64.decode(keyBase64)),
      nonce: Uint8List.fromList(base64.decode(nonceBase64)),
      aad: Uint8List(0),
    );
  }

  Uint8List _aesGcm({
    required bool encrypt,
    required Uint8List data,
    required Uint8List key,
    required Uint8List nonce,
    required Uint8List aad,
  }) {
    final cipher = GCMBlockCipher(AESEngine())
      ..init(encrypt, AEADParameters(KeyParameter(key), 128, nonce, aad));
    return cipher.process(data);
  }

  Uint8List _wrapContentKey(RSAPublicKey key, Uint8List contentKey) {
    final cipher = OAEPEncoding.withSHA256(RSAEngine())
      ..mgf1Hash = SHA1Digest();
    cipher.init(
      true,
      ParametersWithRandom(
        PublicKeyParameter<RSAPublicKey>(key),
        PortableIdentityStore.secureRandom(),
      ),
    );
    return cipher.process(contentKey);
  }

  Uint8List _unwrapContentKey(RSAPrivateKey key, Uint8List wrappedKey) {
    final cipher = OAEPEncoding.withSHA256(RSAEngine())
      ..mgf1Hash = SHA1Digest();
    cipher.init(false, PrivateKeyParameter<RSAPrivateKey>(key));
    return cipher.process(wrappedKey);
  }
}
